<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подключаемся-к-базе-данных-и-знакомимся-с-данными" data-toc-modified-id="Подключаемся-к-базе-данных-и-знакомимся-с-данными-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подключаемся к базе данных и знакомимся с данными</a></span></li><li><span><a href="#Задание-1" data-toc-modified-id="Задание-1-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Задание 1</a></span></li><li><span><a href="#Задание-2" data-toc-modified-id="Задание-2-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Задание 2</a></span></li><li><span><a href="#Задание-3" data-toc-modified-id="Задание-3-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Задание 3</a></span></li><li><span><a href="#Задание-4" data-toc-modified-id="Задание-4-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Задание 4</a></span></li><li><span><a href="#Задание-5" data-toc-modified-id="Задание-5-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Задание 5</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Выводы</a></span></li></ul></div>

# Проект по SQL

Компания решила быть на волне и купила крупный сервис для чтения книг по подписке. 

**Задача** — проанализировать базу данных, в которой содержится информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Описание данных**

**Таблица books**

- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

**Таблица authors**
- author_id — идентификатор автора;
- author — имя автора.

**Таблица publishers**
- publisher_id — идентификатор издательства;
- publisher — название издательства;

**Таблица ratings**
- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

**Таблица reviews**
- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя пользователя, написавшего обзор;
- text — текст обзора.

**Задания**
- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Подключаемся к базе данных и знакомимся с данными

In [1]:
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                  db_config['pwd'],
                                                  db_config['host'],
                                                  db_config['port'],
                                                  db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [2]:
query_1 = '''
            SELECT * 
            FROM books
            LIMIT 5;
'''

In [3]:
query_2 = '''
            SELECT * 
            FROM authors
            LIMIT 5;
'''

In [4]:
query_3 = '''
            SELECT * 
            FROM publishers
            LIMIT 5;
'''

In [5]:
query_4 = '''
            SELECT * 
            FROM ratings
            LIMIT 5;
'''

In [6]:
query_5 = '''
            SELECT * 
            FROM reviews
            LIMIT 5;
'''

In [7]:
pd.io.sql.read_sql(query_1, con = engine) 

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [8]:
pd.io.sql.read_sql(query_2, con = engine) 

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [9]:
pd.io.sql.read_sql(query_3, con = engine) 

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [10]:
pd.io.sql.read_sql(query_4, con = engine) 

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [11]:
pd.io.sql.read_sql(query_5, con = engine) 

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

## Задание 1 

Посчитайте, сколько книг вышло после 1 января 2000 года

In [17]:
query_t1 = '''
           SELECT COUNT(book_id)
           FROM books
           WHERE publication_date > DATE('2000-01-01');

'''

In [16]:
pd.io.sql.read_sql(query_extra2, con = engine) 

count
0      3

*Дополнительные запросы для выводов и анализа базы*

In [14]:
query_extra1 = '''
SELECT MIN(publication_date), MAX(publication_date)
FROM books;
'''

query_extra2 = '''
SELECT COUNT(title)
FROM books
WHERE publication_date > DATE('2015-01-01');
'''

In [15]:
pd.io.sql.read_sql(query_extra1, con = engine) 

min         max
0  1952-12-01  2020-03-31

**Вывод:** в базе присутствуют данные о книгах, изданных **с 1952 года по 2020**. С 2000 года было выпущено более 800 книг (818). Важно, что в последние годы издание книг значительно замедилось: с 2015 до 2020 по данным базы издано всего 3 книги. 

## Задание 2

Для каждой книги посчитайте количество обзоров и среднюю оценку

In [23]:
query_t2 = '''
           SELECT b.title, COUNT(r.review_id) as "количество обзоров", AVG(rat.rating) as "средняя оценка" 
           FROM books b
           LEFT JOIN reviews r ON b.book_id = r.book_id
           LEFT JOIN ratings rat ON b.book_id = rat.book_id
           GROUP BY b.book_id
           ORDER BY COUNT(r.review_id) DESC;

'''

In [20]:
pd.io.sql.read_sql(query_t2, con = engine) 

title  количество обзоров  \
0            The Body in the Library (Miss Marple  #3)                   4   
1                                            Galápagos                   4   
2                             A Tree Grows in Brooklyn                  60   
3    Undaunted Courage: The Pioneering First Missio...                   4   
4                                          The Prophet                  28   
..                                                 ...                 ...   
995                                Alice in Wonderland                  52   
996          A Woman of Substance (Emma Harte Saga #1)                   4   
997                                          Christine                  21   
998  The Magicians' Guild (Black Magician Trilogy  #1)                   4   
999                           The Plot Against America                   4   

     средняя оценка  
0          4.500000  
1          4.500000  
2          4.250000  
3          4.000000  
4          4.285714  
..              ...  
995        4.230769  
996        5.000000  
997        3.428571  
998        3.500000  
999        3.000000  

[1000 rows x 3 columns]

In [21]:
query_books = '''
SELECT COUNT(title)
FROM books
'''

In [22]:
pd.io.sql.read_sql(query_books, con = engine) 

count
0   1000

In [24]:
pd.io.sql.read_sql(query_t2, con = engine) 

title  количество обзоров  \
0                              Twilight (Twilight  #1)                1120   
1                  The Hobbit  or There and Back Again                 528   
2                               The Catcher in the Rye                 516   
3    Harry Potter and the Prisoner of Azkaban (Harr...                 492   
4    Harry Potter and the Chamber of Secrets (Harry...                 480   
..                                                 ...                 ...   
995                          Essential Tales and Poems                   0   
996                            The Natural Way to Draw                   0   
997  Disney's Beauty and the Beast (A Little Golden...                   0   
998    Anne Rice's The Vampire Lestat: A Graphic Novel                   0   
999   The Cat in the Hat and Other Dr. Seuss Favorites                   0   

     средняя оценка  
0          3.662500  
1          4.125000  
2          3.825581  
3          4.414634  
4          4.287500  
..              ...  
995        4.000000  
996        3.000000  
997        4.000000  
998        3.666667  
999        5.000000  

[1000 rows x 3 columns]

In [25]:
query_t2_5 = '''
           SELECT * 
           FROM books b
           LEFT JOIN reviews r ON b.book_id = r.book_id
           LEFT JOIN ratings rat ON b.book_id = rat.book_id
'''
pd.io.sql.read_sql(query_t2_5, con = engine) 

book_id  author_id                                              title  \
0            1        546                                       'Salem's Lot   
1            1        546                                       'Salem's Lot   
2            1        546                                       'Salem's Lot   
3            1        546                                       'Salem's Lot   
4            1        546                                       'Salem's Lot   
...        ...        ...                                                ...   
26175     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   
26176     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   
26177     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   
26178     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   
26179     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

       num_pages publication_date  publisher_id  review_id  book_id  \
0            594       2005-11-01            93        2.0      1.0   
1            594       2005-11-01            93        1.0      1.0   
2            594       2005-11-01            93        2.0      1.0   
3            594       2005-11-01            93        1.0      1.0   
4            594       2005-11-01            93        2.0      1.0   
...          ...              ...           ...        ...      ...   
26175        540       2006-04-25           143     2790.0   1000.0   
26176        540       2006-04-25           143     2793.0   1000.0   
26177        540       2006-04-25           143     2792.0   1000.0   
26178        540       2006-04-25           143     2791.0   1000.0   
26179        540       2006-04-25           143     2790.0   1000.0   

             username                                               text  \
0          ryanfranco  Foot glass pretty audience hit themselves. Amo...   
1        brandtandrea  Mention society tell send professor analysis. ...   
2          ryanfranco  Foot glass pretty audience hit themselves. Amo...   
3        brandtandrea  Mention society tell send professor analysis. ...   
4          ryanfranco  Foot glass pretty audience hit themselves. Amo...   
...               ...                                                ...   
26175         wknight   Change lose answer close pressure. Spend so now.   
26176      jarvispaul  Republican staff bit eat material measure plan...   
26177         wendy18  Or western offer wonder ask. More hear phone f...   
26178  carolrodriguez  Authority go who television entire hair guy po...   
26179         wknight   Change lose answer close pressure. Spend so now.   

       rating_id  book_id       username  rating  
0              1        1     ryanfranco       4  
1              1        1     ryanfranco       4  
2              2        1  grantpatricia       2  
3              2        1  grantpatricia       2  
4              3        1   brandtandrea       5  
...          ...      ...            ...     ...  
26175       6455     1000          zross       2  
26176       6456     1000        fharris       5  
26177       6456     1000        fharris       5  
26178       6456     1000        fharris       5  
26179       6456     1000        fharris       5  

[26180 rows x 14 columns]

In [26]:
query_t2_1 = '''
           SELECT b.title, COUNT(DISTINCT r.review_id) as "количество обзоров", AVG(rat.rating) as "средняя оценка" 
           FROM books b
           LEFT JOIN reviews r ON b.book_id = r.book_id
           LEFT JOIN ratings rat ON b.book_id = rat.book_id
           GROUP BY b.book_id
           ORDER BY COUNT(DISTINCT r.review_id) DESC;

'''

In [27]:
pd.io.sql.read_sql(query_t2_1, con = engine) 

title  количество обзоров  \
0                              Twilight (Twilight  #1)                   7   
1                                  Water for Elephants                   6   
2                                     The Glass Castle                   6   
3    Harry Potter and the Prisoner of Azkaban (Harr...                   6   
4    The Curious Incident of the Dog in the Night-Time                   6   
..                                                 ...                 ...   
995    Anne Rice's The Vampire Lestat: A Graphic Novel                   0   
996                            The Natural Way to Draw                   0   
997   The Cat in the Hat and Other Dr. Seuss Favorites                   0   
998                          Essential Tales and Poems                   0   
999  Disney's Beauty and the Beast (A Little Golden...                   0   

     средняя оценка  
0          3.662500  
1          3.977273  
2          4.206897  
3          4.414634  
4          4.081081  
..              ...  
995        3.666667  
996        3.000000  
997        5.000000  
998        4.000000  
999        4.000000  

[1000 rows x 3 columns]

*Дополнительные запросы для анализа*

In [28]:
query_extra3 = '''
           SELECT b.title, COUNT(DISTINCT r.review_id) as "количество обзоров", AVG(rat.rating) as "средняя оценка", avg(b.num_pages)
           FROM books b
           left JOIN reviews r ON b.book_id = r.book_id
           left JOIN ratings rat ON b.book_id = rat.book_id
           GROUP BY b.book_id
           order by COUNT( DISTINCT r.review_id) desc
           LIMIT 5;
'''

query_extra4 = '''
           SELECT b.title, COUNT(DISTINCT r.review_id) as "количество обзоров", AVG(rat.rating) as "средняя оценка", avg(b.num_pages)
           FROM books b
           left JOIN reviews r ON b.book_id = r.book_id
           left JOIN ratings rat ON b.book_id = rat.book_id
           GROUP BY b.book_id
           order by AVG(rat.rating) desc
           LIMIT 5;
'''

In [29]:
pd.io.sql.read_sql(query_extra3, con = engine) 

title  количество обзоров  \
0                            Twilight (Twilight  #1)                   7   
1                          Outlander (Outlander  #1)                   6   
2                                    Eat  Pray  Love                   6   
3  Harry Potter and the Chamber of Secrets (Harry...                   6   
4  Harry Potter and the Prisoner of Azkaban (Harr...                   6   

   средняя оценка    avg  
0        3.662500  501.0  
1        4.125000  850.0  
2        3.395833  368.0  
3        4.287500  341.0  
4        4.414634  435.0

In [30]:
pd.io.sql.read_sql(query_extra4, con = engine) 

title  количество обзоров  \
0  A Woman of Substance (Emma Harte Saga #1)                   2   
1            Act of Treason (Mitch Rapp  #9)                   2   
2              A Dirty Job (Grim Reaper  #1)                   4   
3      A Fistful of Charms (The Hollows  #4)                   2   
4                              Alas  Babylon                   2   

   средняя оценка    avg  
0             5.0  928.0  
1             5.0  432.0  
2             5.0  387.0  
3             5.0  510.0  
4             5.0  323.0

**Вывод**: у книг с наименьшим количество обзоров наивысшая оценку, в то время как у мировых бестселлеров с огромным количеством обзоров оценки в среднем ниже и ничего не говорит о качестве книги. Это вполне может быть объяснимо: книги с небольшим количеством обзоров могут быть недавно вышедшими (а соответственно, оценок у них в целом меньше, за счет чего оценка выше (если одна 5, то и средняя будет 5), в то время как у давно вышедших, популярных книг много разных оценок, как и много отзывов. Или же книги с маленьким количеством обзоров непопулярны, ограничены, направлены на узкий круг читателей.

## Задание 3

Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [31]:
query_t3 = '''
           SELECT p.publisher
           FROM publishers p
           JOIN books b ON p.publisher_id = b.publisher_id
           WHERE p.publisher_id = (SELECT publisher_id
                                   FROM books
                                   WHERE num_pages > 50
                                   GROUP BY publisher_id
                                   ORDER BY COUNT(book_id) DESC
                                   LIMIT 1)
           LIMIT 1;
'''

In [32]:
pd.io.sql.read_sql(query_t3, con = engine) 

publisher
0  Penguin Books

*Дополнительные запросы для анализа*

In [33]:
query_extra5 = '''
           SELECT p.publisher, count(b.title)
           FROM publishers p
           JOIN books b ON p.publisher_id = b.publisher_id
           WHERE p.publisher_id = (SELECT publisher_id
                                   FROM books
                                   WHERE num_pages > 50
                                   GROUP BY publisher_id
                                   ORDER BY COUNT(book_id) DESC
                                   LIMIT 1)
            GROUP BY p.publisher
            LIMIT 1;
'''

In [34]:
pd.io.sql.read_sql(query_extra5, con = engine) 

publisher  count
0  Penguin Books     42

**Вывод:** издательство, которое выпустило наибольшее число книг толще 50 страниц - Penguin Books. Оно выпустило 42 книги

## Задание 4 

Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [35]:
query_t4 = '''
           SELECT a.author, AVG(r.rating)
           FROM authors a
           JOIN books b ON a.author_id = b.author_id
           JOIN ratings r ON b.book_id = r.book_id
           WHERE b.book_id IN (SELECT r.book_id
                               FROM ratings r
                               GROUP BY r.book_id
                               HAVING COUNT(r.rating_id) > 50)
           GROUP BY a.author
           ORDER BY AVG(r.rating) DESC
           LIMIT 5;

'''

In [36]:
pd.io.sql.read_sql(query_t4, con = engine) 

author       avg
0         J.K. Rowling/Mary GrandPré  4.287097
1  Markus Zusak/Cao Xuân Việt Khương  4.264151
2                     J.R.R. Tolkien  4.246914
3                  Louisa May Alcott  4.192308
4                       Rick Riordan  4.080645

**Вывод:** автор с самой высокой оценкой книг - Джоан Роулинг, а средняя оценка ее книг составляет 4.29. В топе авторов также находятся Маркус Зусак (со средней оценкой 4.26) и Толкиен - 4.25. Средние оценки по книгам у этих авторов очень близки. 

## Задание 5

Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [37]:
query_t5 = '''
           SELECT AVG(mean_number)
           FROM (SELECT count(review_id) as "mean_number" 
                 FROM reviews
                 WHERE username IN (SELECT username
                                   FROM ratings
                                   GROUP BY username
                                   HAVING COUNT(rating) > 50)
                 GROUP BY username) as means;
'''

In [38]:
pd.io.sql.read_sql(query_t5, con = engine) 

avg
0  24.333333

*Дополнительные запросы для анализа*

In [39]:
query_extra6 = '''
SELECT AVG(number)
FROM (SELECT COUNT(review_id) AS "number"
      FROM reviews
      GROUP BY username) AS numbers;
'''

query_extra7 = '''
SELECT AVG(number)
FROM (SELECT COUNT(rating_id) AS "number"
      FROM ratings
      GROUP BY username) AS numbers_1
'''

In [40]:
pd.io.sql.read_sql(query_extra6, con = engine) 

avg
0  17.45625

In [41]:
pd.io.sql.read_sql(query_extra7, con = engine) 

avg
0  40.35

**Вывод:** среднее количество обзоров от пользователей, которые поставили больше 50 оценок, составляет 24. В среднем пользователи оставляют по 17 обзоров и ставят по 40 рейтинговых оценок. 

## Выводы

- в базе присутствуют данные о книгах, изданных с 1952 года по 2020. С 2000 года было выпущено более 800 книг (819). Важно, что в последние годы издание книг значительно замедилось: с 2015 до 2020 по данным базы издано всего 3 книги
- у книг с наименьшим количество обзоров наивысшая оценку, в то время как у мировых бестселлеров с огромным количеством обзоров оценки в среднем ниже и ничего не говорит о качестве книги. Это вполне может быть объяснимо: книги с небольшим количеством обзоров могут быть недавно вышедшими (а соответственно, оценок у них в целом меньше, за счет чего оценка выше (если одна 5, то и средняя будет 5), в то время как у давно вышедших, популярных книг много разных оценок, как и много отзывов. Или же книги с маленьким количеством обзоров непопулярны, ограничены, направлены на узкий круг читателей
- издательство, которое выпустило наибольшее число книг толще 50 страниц - Penguin Books. Оно выпустило 42 книги
- автор с самой высокой оценкой книг - Джоан Роулинг, а средняя оценка ее книг составляет 4.29. В топе авторов также находятся Маркус Зусак (со средней оценкой 4.26) и Толкиен - 4.25. Средние оценки по книгам у этих авторов очень близки
- среднее количество обзоров от пользователей, которые поставили больше 50 оценок, составляет 24. В среднем пользователи оставляют по 17 обзоров и ставят по 40 рейтинговых оценок.